contracts/fate.md

```diff
@@ -0,0 +1,770 @@
+exception is thrown.
+
+The key is a an integer (32 bit).
+
+#### The accumulator stack
+
+The accumulator stack is a stack of typed values. The top of the stack
+can be used as an argument to an operation. Values can be pushed to
+the stack and popped from the stack.
+
+#### The execution stack
+
+The execution stack contain return addresses as a tuple in the form of
+{contract address, function hash, and basic block index}.
+
+#### The event stream
+
+A contract can emit events similar to the EVM.
+
+## Types
+
+The machine supports the following types:
+* Integer (signed arbitrary precision integers) [I]
+* Boolean (true | false) [B]
+* Addresses (A pointer into the state tree) [A]
+* Strings (utf8 encoded byte arrays) [S]
+* Tuples (), ('a), ('a, 'b, ...) [T]
+* Lists: Cons cells (‘a) | Nil (Ʇ)  [L],[C],[Ʇ]
+* Maps (‘a, ‘b),  (Key: ‘a -> Value: ‘b)  ‘a not a map [M]
+* Variant Types, (| Size | Tag | Elements  |)  [V]
+* Bits <Bits> or !<Bits>
+* TypeRep [R]
+
+These types can be used as argument and return types of functions and
+also as the type of the elements in the contract store.
+
+### Integer
+
+The type Integer, is used for any integer, infinitely large or small
+(below zero).
+
+Examples:
+
+```
+ 0
+ 1
+-1
+589465789334
+-51315353513513131656462262
+```
+
+### Boolean
+
+The Boolean type only has two values, true or false.
+
+Examples (all of them):
+```
+ true
+ false
+```
+
+Operations on booleans include:
+*   and, or, not
+*   conditional jumps
+
+### Addresses
+
+Values of the type Address are 32-bytes (256-bits) pointers to
+entities on the chain (Accounts, Oracles, Contracts, etc)
+
+Examples:
+```
+<<ak_2HNb8bivUoJTcaxD6VKDy2wPHvTuQgQRJ3dkF9RSyVFqBkMneC>>
+```
+
+### Strings
+
+Strings are byte arrays of UTF-8 encoded unicode characters. (These
+can be used as non-indexable arrays of bytes.)
+
+Examples:
+```
+“Hello world”
+  “eof”
+```
+
+### Tuples
+
+Tuples have an arity indicating the number of elements in the
+tuple. Each element in the tuple has a type. The 0-tuple is also
+sometimes called unit.
+
+Examples:
+```
+ {}                Type: unit
+ {1, 2}            Type: Tuple(Integer, Integer)
+ {“foo”, 42, true} Type: Tuple(String, Integer, Boolean)
+```
+
+### Lists
+
+Lists are monomorphic series of values. Hence lists have a type. There
+is also an empty list (or Nil).
+
+Examples:
+```
+[]                   Type: Nil
+[1, 2, 3]            Type: List(Integer)
+[true, true, false]  Type: List(Boolean)
+```
+
+### Maps
+
+Maps are monomorphic key value stores, that is the key is always the
+same type for all elements in a map and the value has the same type in
+all mappings in a map.  The key can have any type except a map.  The
+value can have any type including a map.
+
+Examples:
+```
+#{ (1, “foo”), (2, “bar”) }
+```
+Type: Map(Integer, String)
+```
+#{ (“Fruit prices”, #{ (“banana”, 42), (“apple”, 12)}),
+   (“Stock prices”, #{(“Apple”, 142), (“Orange”, 112)}) }
+```
```

Comment:
Add the type for the second one as well?

Reply:
`Type: Map(String, Map(String, Integer))`